### PR TITLE
BVH/ImplicitGrid API refactor

### PR DESCRIPTION
**Pull request by kanye-quest**
# Summary

- Refactors `spin::BVH` public API and internal implementation to make use of `axom::Array`/`axom::ArrayView`
  - Public `BVH::findPoints/findRays/findBoundingBoxes()` methods now accept `offsets` and `counts` as `ArrayView`s, and output `candidates` as an `axom::Array`
  - Internal BVH data is stored as `axom::Array`s, which Traverser object now points to with `axom::ArrayView`s
- Cleans up `spin::ImplicitGrid` API use of `Array`/`ArrayView`
  - `spin::ImplicitGrid::getCandidatesAsArray()` method now accepts `offsets` and `counts` as externally-allocated `ArrayView`s
  - Adds overloads for `spin::ImplicitGrid::getCandidatesAsArray()` which accept the array of query primitives as an `ArrayView`
  - Use `ArrayView` to replace raw pointers in `ImplicitGrid::QueryObject`
- Fixes a bug with `mesh_tester` binary on blueos by enabling unified memory

### CODE REVIEW

**Review by kennyweiss — APPROVED** [review]
Thanks @kanye-quest -- this looks much cleaner!

Please document the API changes for `BVH` and `ImplicitGrid` in the release notes and user docs (if applicable).

File: src/axom/spin/internal/linear_bvh/build_radix_tree.hpp [hunk 1]
```diff
@@ -525,6 +519,8 @@ void propagate_aabbs(RadixTree<FloatType, NDIMS>& data, int allocatorID)
 {
   AXOM_PERF_MARK_FUNCTION("propagate_abbs");
 
+  using BoxType = primal::BoundingBox<FloatType, NDIMS>;
```
Comment (agcapps) [inline src/axom/spin/internal/linear_bvh/build_radix_tree.hpp:522]: 👍

File: src/axom/spin/internal/linear_bvh/build_radix_tree.hpp [hunk 2]
```diff
@@ -533,24 +529,23 @@ void propagate_aabbs(RadixTree<FloatType, NDIMS>& data, int allocatorID)
...
-  using BoxType = primal::BoundingBox<FloatType, NDIMS>;
+  const auto inner_aabb_ptr = data.m_inner_aabbs.view();
+  for_all<ExecSpace>(
+    inner_size,
+    AXOM_LAMBDA(IndexType idx) { inner_aabb_ptr[idx] = BoxType {}; });
 
-  array_memset<ExecSpace>(counters_ptr, inner_size, 0);
-  array_memset<ExecSpace>(inner_aabb_ptr, inner_size, BoxType {});
+  Array<int32> counters(inner_size, inner_size, allocatorID);
+  const auto counters_ptr = counters.view();
```
Comment (agcapps) [inline src/axom/spin/internal/linear_bvh/build_radix_tree.hpp:543]: Please fix odd indentation.

File: src/axom/spin/internal/linear_bvh/bvh_vtkio.hpp [hunk 3]
```diff
@@ -152,8 +152,8 @@ void write_root(const primal::BoundingBox<FloatType, 3>& root,
 
 //------------------------------------------------------------------------------
 template <typename FloatType, int NDIMS>
-void write_recursive(primal::BoundingBox<FloatType, NDIMS>* inner_nodes,
-                     int32* inner_node_children,
+void write_recursive(const primal::BoundingBox<FloatType, NDIMS>* inner_nodes,
+                     const int32* inner_node_children,
```
Comment (agcapps) [inline src/axom/spin/internal/linear_bvh/bvh_vtkio.hpp:156]: pointers here, not ArrayView?

**Review by agcapps — APPROVED** [review]
@kanye-quest , thanks very much!  I think I only had a few style queries.

**Comment by kanye-quest** [comment]
I noted the change in the release notes for `BVH` -- since `ImplicitGrid`'s vector API hasn't been in a release yet, I didn't document that change.

**Pull request merged**